blog/_posts/2019-09-12-tips-on-moving-your-dev-env-from-docker-compose-to-kubernetes.md

```diff
@@ -0,0 +1,173 @@
+---
+slug: tips-on-moving-your-dev-env-from-docker-compose-to-kubernetes
+date: 2019-09-12
+author: nick
+layout: blog
+title: "Tips on Moving your Dev Env from Docker-Compose to Kubernetes"
+image: olivier-miche-Skva4bc_Bsg-unsplash.png
+image_caption: "Photo by Olivier Miche on <a href='https://unsplash.com/photos/Skva4bc_Bsg'>Unsplash</a>"
+tags:
+  - kubernetes
+  - docker-compose
+  - developer-tools
+keywords:
+  - kubernetes
+  - docker-compose
+  - developer-tools
+---
+
+When I first started learning how to write Kubernetes configs, I would sometimes
+complain to people about it. "They're so complicated!" they would complain back.
+
+They would show me an example. Here's a simple Docker-Compose config:
+
+```yaml
+app:
+  image: tilt.dev/simple-node-app
+  ports:
+    - 8000:8000
+  command: sh -c 'node server.js'
+```
+
+And here's the equivalent Kubernetes config:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: simple-node-app
+  name: simple-node-app
+spec:
+  selector:
+    matchLabels:
+      app: simple-node-app
+  template:
+    metadata:
+      labels:
+        app: simple-node-app
+    spec:
+      containers:
+      - command: ["sh", "-c", "node server.js"]
+        image: tilt.dev/simple-node-app
+        name: app
+        ports:
+        - containerPort: 8000
+```
+
+"Why can't Kubernetes configs be as simple as my Docker-Compose configs?"
+
+After two years of playing around with both of them, I think I've figured out
+the answer!!
+
+---
+
+Kubernetes configs force us to handle change.
+
+When I write the Docker-Compose file, I get to optimize for a world where:
+
+- Our servers only run on a single machine
+- One instance of each server runs at a time
+- When I reload, I stop all the servers and start them all up again
+
+When I write a Kubernetes file, I have to think about how to upgrade that system:
+
+- Our servers could run on different machines
+- Incremental updates are safe
+- Each component is its own file. If we make a mess, we can swap out one component at a time.
+
+Short Docker-Compose files are survivorship bias in action. They have a much
+harder time handling incremental change.
```

Comment:
As in, Docker Compose files have a harder time handling incremental changes?

Maybe just me, but imo this sentence could use more dot-connecting. try:

> We see succinct, elegant Docker Compose files everywhere, but that doesn't mean that _all_ Docker Compose files are like that. In fact, it's survivorship bias in action: we don't see the big Docker Compose files because they're gnarly, brittle messes.

Reply:
done